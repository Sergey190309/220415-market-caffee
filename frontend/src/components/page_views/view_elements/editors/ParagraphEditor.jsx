import React, {
  Fragment,
  useState, useEffect,
  useRef
} from 'react'
import PropTypes from 'prop-types'

import { Form, Segment } from 'semantic-ui-react'
import TextareaAutosize from 'react-textarea-autosize'

import EditorContextMenu from './EditorContextMenu'
// import { contentSaga } from '../../../../redux/saga/contentLoading/contentLoading'
import { createContextFromEvent } from './createContextFromEvent'
import { warningColor } from '../../../../utils/colors'

const ParagraphEditor = ({
  setParagraphEditted,
  comingContent,
  setComimgContent
}) => {
  const [content, setContent] = useState({
    title: '',
    content: '' // conferted from [''] that on site
  })
  const [editorContextMenuOpened, setEditorContextMenuOpened] = useState(false)
  // const [contextMenuOpened, setContextMenuOpened] = useState(false)

  const contextRef = useRef(null)
  useEffect(() => {
    // console.log('ParagraphEditor, useEffect, comingContent ->', comingContent)
    const _text = comingContent.content.join('\n')
    const _content = { ...comingContent, content: _text }
    setContent(_content)
  }, [comingContent])

  const contextMenuAction = (action) => {
    // console.log('ParagraphEditor, contextMenuAction, action ->', action)
    if (action === 'save') {
      const _text = content.content.split('\n')
      const _content = { ...content, content: _text }
      setComimgContent(_content)
      setParagraphEditted(false)
    }
    if (action === 'cancel') {
      setParagraphEditted(false)
    }
  }

  const onChangeHandling = event => {
    // const onChangeHandling = (event, { name }) => {
    // console.log('ParagraphEditor, onChangeHandling, event.target ->', event.target.name)
    event.preventDefault()
    const _content = { ...content, [event.target.name]: event.target.value }
    setContent(_content)
  }

  const onContextMenuHendler = event => {
    event.preventDefault()
    contextRef.current = createContextFromEvent(event)
    setEditorContextMenuOpened(true)
  }

  return (
    <Fragment>
      <EditorContextMenu
        isOpened={editorContextMenuOpened}
        context={contextRef}
        setContextMenuOpened={setEditorContextMenuOpened}
        contextMenuAction={contextMenuAction}
      />
      <Segment
        style={{ backgroundColor: warningColor }}
        onContextMenu={onContextMenuHendler}
      >
        <Form>
          <TextareaAutosize
            name='title'
            value={content.title}
            autoFocus
            rows='1'
            onChange={onChangeHandling}
            style={{
              fontSize: '1.15em',
              fontWeight: 'bold'
            }}
          />
          <TextareaAutosize
            name='content'
            value={content.content}
            rows={
              content.content ? content.content.length : 1
            }
            onChange={onChangeHandling}
          />
        </Form>
      </Segment>

    </Fragment>
  )
}

ParagraphEditor.defaultProps = {
  setParagraphEditted: () => { },
  comingContent: {
    title: '',
    content: ['']
  },
  setComimgContent: () => {}
}

ParagraphEditor.propTypes = {
  setParagraphEditted: PropTypes.func.isRequired,
  comingContent: PropTypes.object.isRequired,
  setComimgContent: PropTypes.func.isRequired
}

export default ParagraphEditor
