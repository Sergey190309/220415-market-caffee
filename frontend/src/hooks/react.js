import {useState, useEffect, useRef, useCallback } from 'react'

export const useAppState = useState
export const useAppEffect = useEffect
export const useAppRef = useRef
export const useAppCallback = useCallback
