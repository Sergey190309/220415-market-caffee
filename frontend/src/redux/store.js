import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit'
import createSagaMiddleware from 'redux-saga'
// import reducer from './reducers';
import reducer from './slices'
import rootSaga from './saga'
// import { rootReducer } from './reducers';

const initialState = {}
const sagaMiddleware = createSagaMiddleware()

const middleware = [...getDefaultMiddleware({ thunk: false }), sagaMiddleware]
// const middleware = [sagaMiddleware, thunk];

const store = configureStore({
  reducer,
  middleware,
  initialState
})

sagaMiddleware.run(rootSaga)

export default store
