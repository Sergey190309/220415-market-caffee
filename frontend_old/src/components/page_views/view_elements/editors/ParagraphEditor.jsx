import React, {
  Fragment,
  useState, useEffect,
  useRef
} from 'react'
import PropTypes from 'prop-types'

import { Form, Segment } from 'semantic-ui-react'
import TextareaAutosize from 'react-textarea-autosize'

import EditorContextMenu from './EditorContextMenu'
import { createContextFromEvent } from '../../../../utils/createContext'
import { warningColor } from '../../../../utils/colors'

const ParagraphEditor = ({
  comingContent,
  setParagraphEditted,
  setComingContent
}) => {
  const [content, setContent] = useState({
    title: '',
    content: '' // conferted from [''] that on site
  })
  const [editorContextMenuOpened, setEditorContextMenuOpened] = useState(false)
  const [changed, setChanged] = useState(false)
  // const [contextMenuOpened, setContextMenuOpened] = useState(false)

  const contextRef = useRef(null)

  useEffect(() => {
    // console.log('ParagraphEditor, useEffect, comingContent ->', comingContent.content)
    const title = comingContent.title ? comingContent.title : ''
    const content = comingContent.content.join('\n')
    setContent({ title, content })
  }, [comingContent])

  const contextMenuAction = (action) => {
    if (action === 'save') {
      // console.log('ParagraphEditor:\n contextMenuAction',
      //   '\n  content ->', content)
      const _text = content.content === ''
        ? []
        : content.content.split('\n')
      const _content = { ...content, content: _text }
      setComingContent(_content)
      setParagraphEditted(false)
    }
    if (action === 'cancel') {
      setParagraphEditted(false)
    }
  }

  const onChangeHandling = event => {
    // const onChangeHandling = (event, { name }) => {
    event.preventDefault()
    const _content = { ...content, [event.target.name]: event.target.value }
    setChanged(true)
    setContent(_content)
  }

  const onContextMenuHendler = event => {
    // console.log('ParagraphEditor, onContextMenuHendler')
    event.preventDefault()
    contextRef.current = createContextFromEvent(event)
    setEditorContextMenuOpened(true)
  }

  return (
    <Fragment>
      <EditorContextMenu
        isOpened={editorContextMenuOpened}
        saveDisabled={!changed}
        context={contextRef}
        setContextMenuOpened={setEditorContextMenuOpened}
        contextMenuAction={contextMenuAction}
      />
      <Segment
        style={{ backgroundColor: warningColor }}
        onContextMenu={onContextMenuHendler}
        data-testid='Segment'
      >
        <Form>
          <TextareaAutosize
            name='title'
            value={content.title}
            autoFocus
            rows={1}
            onChange={onChangeHandling}
            style={{
              fontSize: '1.15em',
              fontWeight: 'bold'
            }}
          />
          <TextareaAutosize
            name='content'
            value={content.content}
            rows={1}
            onChange={onChangeHandling}
          />
        </Form>
      </Segment>

    </Fragment>
  )
}

ParagraphEditor.defaultProps = {
  comingContent: {
    title: '',
    content: ['']
  },
  setParagraphEditted: () => { },
  setComingContent: () => {}
}

ParagraphEditor.propTypes = {
  comingContent: PropTypes.object.isRequired,
  setParagraphEditted: PropTypes.func.isRequired,
  setComingContent: PropTypes.func.isRequired
}

export default ParagraphEditor
