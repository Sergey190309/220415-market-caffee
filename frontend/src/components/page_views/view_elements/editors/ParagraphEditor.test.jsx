/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
import React from 'react'
// import { Provider } from 'react-redux'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { testRender } from '../../../../tstHelpers'

import store from '../../../../redux/store'

import ParagraphEditor from './ParagraphEditor'
// import EditorContextMenu from './EditorContextMenu'

const mockEditorContextMenu = jest.fn()

jest.mock('./EditorContextMenu', () => props => {
  mockEditorContextMenu(props)
  return <div data-testid='mockEditorContextMenu' />
  // return <mock-editorContextMenu />
})

describe('ParagraphEditor', () => {
  beforeEach(() => {
    // EditorContextMenu.mockImplementation(MockEditorContextMenu)
  })
  const testProps = {
    comingContent: {
      title: 'mockTitle',
      content: ['mockContent first paragraph', 'second paragraph', 'third one']
    },
    setParagraphEditted: jest.fn(),
    setComimgContent: jest.fn()
  }
  describe('appeance', () => {
    test('rendering (snapshot)', async () => {
      testRender(<ParagraphEditor {...testProps} />, { store: store })
      const segment = await screen.findByTestId('Segment')
      expect(segment).toMatchSnapshot()
      // screen.debug()
    })
  })
  describe('actions', () => {
    it('should contain proper values', () => {
      testRender(<ParagraphEditor {...testProps} />, { store: store })
      const textAreas = screen.getAllByRole('textbox')
      expect(textAreas).toHaveLength(2)
      expect(textAreas[0]).toHaveValue(testProps.comingContent.title)
      expect(textAreas[1]).toHaveValue(testProps.comingContent.content.join('\n'))

      // screen.debug()
    })
    it('text areas user input', () => {
      testRender(<ParagraphEditor {...testProps} />, { store: store })
      const titleInput = 'Test value'
      const titleTextArea = screen.getAllByRole('textbox')[0]
      userEvent.clear(titleTextArea)
      userEvent.type(titleTextArea, titleInput)
      expect(titleTextArea).toHaveValue(titleInput)

      const contentInput = 'first{enter}second{enter}third'
      const expContentInput = contentInput.split('{enter}').join('\n')
      const contentTextArea = screen.getAllByRole('textbox')[1]
      userEvent.clear(contentTextArea)
      userEvent.type(contentTextArea, contentInput)
      expect(contentTextArea).toHaveValue(expContentInput)
    })
    it('right click -> context menu with appropriate args', () => {
      testRender(<ParagraphEditor {...testProps} />, { store: store })
      const element = screen.getByTestId('Segment')
      userEvent.click(element, { button: 2 })
      const args = mockEditorContextMenu.mock.calls[mockEditorContextMenu.mock.calls.length - 1][0]
      expect(args.isOpened).toBeTruthy()
      expect(args.saveDisabled).toBeTruthy()
      expect(args.context).toMatchObject({ current: {} })
      expect(args.setContextMenuOpened).toEqual(expect.any(Function))
      expect(args.contextMenuAction).toEqual(expect.any(Function))
      // console.log('EditorContextMenu, args.setContextMenuOpened ->',
      //   args.setContextMenuOpened)
    })
  })
})
