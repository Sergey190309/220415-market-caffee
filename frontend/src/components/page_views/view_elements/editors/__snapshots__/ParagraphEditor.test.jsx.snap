// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParagraphEditor appeance rendering (snapshot) 1`] = `
<div
  class="ui segment"
  data-testid="Segment"
  style="background-color: orange;"
>
  <form
    class="ui form"
  >
    <textarea
      name="title"
      rows="1"
      style="font-size: 1.15em; font-weight: bold;"
    >
      mockTitle
    </textarea>
    <textarea
      name="content"
      rows="1"
    >
      mockContent first paragraph
second paragraph
third one
    </textarea>
  </form>
</div>
`;
